share/smack/top.py

```diff
@@ -899,85 +979,133 @@ def verify_bpl(args):
         command += args.verifier_options.split()
 
     verifier_output = try_command(command, timeout=args.time_limit)
-    verifier_output = transform_out(args, verifier_output)
-    result = verification_result(verifier_output, args.verifier)
+    return process_verifier_output(args, verifier_output)
 
-    if args.json_file:
-        with open(args.json_file, 'w') as f:
-            f.write(json_output_str(result, verifier_output, args.verifier))
 
-    if result in VResult.ERROR:
-        error = error_trace(verifier_output, args.verifier)
+def get_result(result):
+    return result
 
-        if args.error_file:
-            with open(args.error_file, 'w') as f:
-                f.write(error)
 
-        if not args.quiet:
-            print(error)
+def thread_verify_bpl(args, commands_to_add):
+    args.modular = False
 
-        if args.replay:
-            replay_error_trace(verifier_output, args)
-    print(result.message(args))
-    return result.return_code()
+    if 'verifier' in commands_to_add:
+        args.verifier = commands_to_add['verifier']
+    else:
+        args.verifier = 'corral'
+
+    if 'modular' in commands_to_add:
+        args.modular = commands_to_add['modular']
+
+    if 'verifier-options' in commands_to_add:
+        args.verifier_options = commands_to_add['verifier-options']
+
+    if 'solver' in commands_to_add:
+        args.solver = commands_to_add['solver']
+
+    if args.verifier == 'boogie' or args.modular:
+        command = invoke_boogie(args)
+
+    elif args.verifier == 'corral':
+        command = invoke_corral(args)
+
+    elif args.verifier == 'symbooglix':
+        command = invoke_symbooglix(args)
+
+    if args.verifier_options:
+        command += args.verifier_options.split()
+
+    return args, try_command(command, timeout=args.time_limit)
+
+
+def verify_bpl_portfolio(args):
+    if args.portfolio_config is None:
+        portfolio_config = {'thread1':
+                               {'verifier': 'corral',
+                                'verifier-options': ""},
+                            'thread2':
+                               {'verifier': 'corral',
+                                'verifier-options':
+                                    "/bopt:proverOpt:O:smt.arith.solver=2"},
+                            'thread3':
+                               {'verifier': 'corral',
+                                'verifier-options':
+                                    "/bopt:proverOpt:O:smt.qi.eager_threshold=100"},
+                            'thread4':
+                               {'verifier': 'corral',
+                                'verifier-options':
+                                   ("/bopt:proverOpt:O:smt.qi.eager_threshold=100"
+                                    " /bopt:proverOpt:O:smt.arith.solver=2")}}
+    else:
+        portfolio_config = yaml.load(open(args.portfolio_config, 'r'))
+
+    p = multiprocessing.Pool()
+    thread_args = copy.deepcopy(args)
```

Comment:
How come you don't have to create a deep copy of args for each thread?
Aren't `thread_args` going to get messed up if you keep overwriting them for each thread?

Reply:
I had the same thought but it did not happen. I will just change this though because it will be easier to understand going forwards if everyone that looks at it has that thought

Reply:
Hmmm, that's weird. I wonder if that is because of how the configurations that you tested with looked like. In many scenarios you would not notice this problem since overwriting will still do the right thing.
Otherwise, I have no reasonable explanation for why the current implementation would work properly.